### PR TITLE
[Metrics] Add multiclass auroc

### LINKED ISSUE #3304: MulticlassAUROC: Implement a multi-class version of the AUROC metric
## 🚀 Feature

Create the metric MulticlassAUROC to allow for the AUROC metric to be used in multi-class problem settings. Or,
Expand the AUROC metric to support multi-class data, which would also directly solve this AUROC bug that instead gives a random value when used in multi-class problems: https://github.com/PyTorchLightning/pytorch-lightning/issues/3303

### Motivation

AUROC is a useful metric for model performance, but the current AUROC metric is not made for multi-class problems.
An out of the box working multi-class AUROC metric would be great.

### Pitch

According to [this post on the PyTorch Lightning forum](https://forums.pytorchlightning.ai/t/pytorch-lightning-auroc-value-for-multi-class-seems-to-be-completely-off-compared-to-sklearn-using-it-wrong/61/7), implementation would just require wrapping existing functionality together.

Implementing this metric would lower the barrier of its usage, instead of needing to be aware of the different metrics available that could be combined to effectively get his metric.

### Alternatives

- Use sklearn's implementation of AUROC: [`sklearn.metrics.roc_auc_score`](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html).
This slows down the training, however, due to the need of transferring prediction and target data to CPU.
- Implement manually by combining [multiclass auc](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/metrics/functional/classification.py#L825) and a [multiclass roc calculation](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/metrics/functional/classification.py#L620)

### Additional context

- Relevant discussion: https://forums.pytorchlightning.ai/t/pytorch-lightning-auroc-value-for-multi-class-seems-to-be-completely-off-compared-to-sklearn-using-it-wrong/61/6
- Relevant AUROC bug when used for multi-class: https://github.com/PyTorchLightning/pytorch-lightning/issues/3303

### PR DESCRIPTION
**Pull request by ddrevicky**
## What does this PR do?

Implements functional multiclass AUROC.

Fixes #3304 

Notes on the code:

Had to pass `reorder=False` for `auc` because tests against `sklearn` kept showing different values and debugging showed that the difference was actually coming out of our `auc` using `torch.argsort` which is [unstable](https://github.com/pytorch/pytorch/issues/38681). A short [colab](https://colab.research.google.com/drive/18bG1cqSf-KJOmvZWKUtGUASZagrDqxf6#scrollTo=S9_kfT5fqBpH) notebook documenting this. I submitted a separate issue #4237 for that.

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/metrics/functional/classification.py [hunk 1]
```diff
@@ -900,6 +901,62 @@ def _auroc(pred, target, sample_weight, pos_label):
...
+            f"the number of classes passed in 'num_classes' ({num_classes}).")
+
+    @multiclass_auc_decorator(reorder=False)
+    def _multiclass_auroc(pred, target, sample_weight, num_classes):
+        return multiclass_roc(pred, target, sample_weight, num_classes)
+
+    class_aurocs = _multiclass_auroc(pred=pred, target=target,
+                                     sample_weight=sample_weight,
+                                     num_classes=num_classes)
+    return torch.mean(class_aurocs)
```
Comment (ddrevicky) [inline pytorch_lightning/metrics/functional/classification.py:957]: I've implemented this using the `multiclass_auc_decorator` similarly as is done for `auroc` but have to say that as not such an experienced Pythonist I was scratching my head for a good while trying to figure out what the `multiclass_auc_decorator` was doing. It's possible that for other people reading the code it might also take unnecessary amount of time. Would the following be more readable? No decorator is needed either outside or within the `multiclass_auroc` function. Just my humble opinion, which do you guys prefer? :)

```
class_rocs = multiclass_roc(pred=pred, target=target, sample_weight=sample_weight, num_classes=num_classes)
class_aurocs = []
for fpr, tpr, _ in class_rocs:
    class_aurocs.append(auc(fpr, tpr, reorder=False))
return torch.mean(torch.stack(class_aurocs))
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -27,6 +27,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 ### Added
 
 - Added trace functionality to the function `to_torchscript` ([#4142](https://github.com/PyTorchLightning/pytorch-lightning/pull/4142))
+- Added multiclass AUROC metric ([#4236](https://github.com/PyTorchLightning/pytorch-lightning/pull/4236))
```
Comment (Borda) [inline CHANGELOG.md:30]:
```suggestion

- Added multiclass AUROC metric ([#4236](https://github.com/PyTorchLightning/pytorch-lightning/pull/4236))
```

File: tests/metrics/functional/test_classification.py [hunk 3]
```diff
@@ -346,6 +347,52 @@ def test_auroc(pred, target, expected):
     assert score == expected
 
 
+def test_multiclass_auroc():
+    with pytest.raises(ValueError,
+                       match="Multiclass AUROC metric expects the target scores to be "
```
Comment (Borda) [inline tests/metrics/functional/test_classification.py:352]: we can catch just a part of the message, no need to have it complete, for example from the recommendation or i.e. section

File: tests/metrics/functional/test_classification.py [hunk 4]
```diff
@@ -346,6 +347,52 @@ def test_auroc(pred, target, expected):
...
+        _ = multiclass_auroc(pred=torch.rand((5, 4)).softmax(dim=1),
+                             target=torch.tensor([0, 1, 2, 2, 3]),
+                             num_classes=6)
+
+
+def test_multiclass_auroc_against_sklearn():
+    device = 'cuda' if torch.cuda.is_available() else 'cpu'
+
+    n_samples = 300
+    for n_cls in [2, 5, 10, 50]:
```
Comment (Borda) [inline tests/metrics/functional/test_classification.py:378]: can we have this in parametrization so it easier to spot the issue...

**Review by teddykoker — APPROVED** [review]
LGTM

File: tests/metrics/functional/test_classification.py (hunk 3 above)
Comment (ddrevicky) [reply tests/metrics/functional/test_classification.py]: Done.

File: tests/metrics/functional/test_classification.py (hunk 4 above)
Comment (ddrevicky) [reply tests/metrics/functional/test_classification.py]: Done.

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/metrics/functional/classification.py [hunk 5]
```diff
@@ -900,6 +901,62 @@ def _auroc(pred, target, sample_weight, pos_label):
...
+        ...                      [0.05, 0.05, 0.85, 0.05],
+        ...                      [0.05, 0.05, 0.05, 0.85]])
+        >>> target = torch.tensor([0, 1, 3, 2])
+        >>> multiclass_auroc(pred, target)   # doctest: +NORMALIZE_WHITESPACE
+        tensor(0.6667)
+    """
+    if not torch.allclose(pred.sum(dim=1), torch.tensor(1.0)):
+        raise ValueError(
+            "Multiclass AUROC metric expects the target scores to be "
+            "probabilities, i.e. they should sum up to 1.0 over classes")
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/classification.py:936]:
```suggestion
        raise ValueError(
            "Multiclass AUROC metric expects the target scores to be"
            " probabilities, i.e. they should sum up to 1.0 over classes")
```

File: pytorch_lightning/metrics/functional/classification.py [hunk 6]
```diff
@@ -900,6 +901,62 @@ def _auroc(pred, target, sample_weight, pos_label):
...
+
+    if torch.unique(target).size(0) != pred.size(1):
+        raise ValueError(
+            f"Number of classes found in in 'target' ({torch.unique(target).size(0)}) "
+            f"does not equal the number of columns in 'pred' ({pred.size(1)}). "
+            "Multiclass AUROC is not defined when all of the classes do not "
+            "occur in the target labels.")
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/classification.py:943]:
```suggestion
            f"Number of classes found in in 'target' ({torch.unique(target).size(0)})"
            f" does not equal the number of columns in 'pred' ({pred.size(1)})."
            " Multiclass AUROC is not defined when all of the classes do not"
            " occur in the target labels.")
```

File: pytorch_lightning/metrics/functional/classification.py [hunk 7]
```diff
@@ -900,6 +901,62 @@ def _auroc(pred, target, sample_weight, pos_label):
...
+
+    if num_classes is not None and num_classes != pred.size(1):
+        raise ValueError(
+            f"Number of classes deduced from 'pred' ({pred.size(1)}) does not equal "
+            f"the number of classes passed in 'num_classes' ({num_classes}).")
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/classification.py:948]:
```suggestion
            f"Number of classes deduced from 'pred' ({pred.size(1)}) does not equal"
            f" the number of classes passed in 'num_classes' ({num_classes}).")
```

**Review by Borda — APPROVED** [review]
lgtm

File: pytorch_lightning/metrics/functional/classification.py [hunk 8]
```diff
@@ -900,6 +901,62 @@ def _auroc(pred, target, sample_weight, pos_label):
...
+        >>> target = torch.tensor([0, 1, 3, 2])
+        >>> multiclass_auroc(pred, target)   # doctest: +NORMALIZE_WHITESPACE
+        tensor(0.6667)
+    """
+    if not torch.allclose(pred.sum(dim=1), torch.tensor(1.0)):
+        raise ValueError(
+            "Multiclass AUROC metric expects the target scores to be"
+            " probabilities, i.e. they should sum up to 1.0 over classes")
+
+    if torch.unique(target).size(0) != pred.size(1):
```
Comment (tchaton) [inline pytorch_lightning/metrics/functional/classification.py:938]: Shouldn't it be `torch.unique(target).size(0) <= pred.size(1)`

```
target = torch.tensor([0, 0, 0, 0])
>>> torch.unique(target).size(0)
1
```

File: pytorch_lightning/metrics/functional/classification.py [hunk 9]
```diff
@@ -900,6 +901,62 @@ def _auroc(pred, target, sample_weight, pos_label):
...
+    Example:
+
+        >>> pred = torch.tensor([[0.85, 0.05, 0.05, 0.05],
+        ...                      [0.05, 0.85, 0.05, 0.05],
+        ...                      [0.05, 0.05, 0.85, 0.05],
+        ...                      [0.05, 0.05, 0.05, 0.85]])
+        >>> target = torch.tensor([0, 1, 3, 2])
+        >>> multiclass_auroc(pred, target)   # doctest: +NORMALIZE_WHITESPACE
+        tensor(0.6667)
+    """
```
Comment (tchaton) [inline pytorch_lightning/metrics/functional/classification.py:932]: Also, we should check pred.size(0) == target.size(0)

File: pytorch_lightning/metrics/functional/classification.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/metrics/functional/classification.py:938]: Or we could have a get_num_classes utils too.

**Review by tchaton — CHANGES_REQUESTED** [review]
Great PR ! Some extra tests on matrix size would be great !

File: pytorch_lightning/metrics/functional/classification.py (hunk 8 above)
Comment (ddrevicky) [reply pytorch_lightning/metrics/functional/classification.py:938]: Well the metric is undefined when `torch.unique(target).size(0) < pred.size(1)`, that's why there is a strict equals. The way this implementation works (based on sklearn's version) is, that it uses a one-vs-rest strategy of computing the AUROC. For `n` classes it computes `n` binary AUROCs, for each class in turn, it is considered a positive class and all the other classes negative. Then it averages those.

E.g., for `n=3` and `target=[0, 1, 1, 2]`, for class `0` we binarize the target to make `0` the positive class: `[1, 0, 0, 0]` and compute the `AUC` of `ROC` of that.

If a target label is not present in the `target`, e.g. `n=3` and `target=[0, 1, 1, 1]` then for the absent class `2` the binarized target would look like `[0, 0, 0, 0]` (all negative) and `ROC` cannot be computed (would raise an error). Consequently, the whole multiclass AUROC is undefined in that case.

File: pytorch_lightning/metrics/functional/classification.py (hunk 9 above)
Comment (ddrevicky) [reply pytorch_lightning/metrics/functional/classification.py:932]: I can add that of course but this is not a check that is done in any other metric implementation. So if it's done here it should probably be done everywhere. If that's desired, I could add a helper to `classification.py`

```
def check_batch_dims(pred, target):
    if not pred.size(0) == target.size(0):
         raise ValueError(f"Batch size for prediction ({pred.size(0)}) and target ({target.size(0)}) must be equal.")

```
Would that work? Then this helper could be used in each metric instead of copy pasting the `if` clause and the exception.

Comment (SkafteNicki) [reply pytorch_lightning/metrics/functional/classification.py:932]: As we are slowly unifying the functional and class based interface, we are doing more checks for shape, so this will come in a future PR :]

File: pytorch_lightning/metrics/functional/classification.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/metrics/functional/classification.py:938]: Make sense !

**Review by tchaton — APPROVED** [review]

File: CHANGELOG.md [hunk 10]
```diff
@@ -106,6 +106,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Added trace functionality to the function `to_torchscript` ([#4142](https://github.com/PyTorchLightning/pytorch-lightning/pull/4142))
 
+- Added multiclass AUROC metric ([#4236](https://github.com/PyTorchLightning/pytorch-lightning/pull/4236))
```
Comment (SkafteNicki) [inline CHANGELOG.md:109]: @ddrevicky could you move this to the unreleased section?
